README.md
# *Kubernetes-prometheus-publisher* service

## Introduction
This service is responsible to:
- retrieve monitoring data related to the containers/pods that are running in a Kubernetes cluster using the Prometheus API,
- publish the monitoring data by given Kubernetes containers/pods in the publish/subscribe broker (Apache Kafka)

Actually, this service feeds the [MAPE](https://github.com/5g-media/mape) with monitoring data.

## Requirements
- Python 3.5+
- Docker engine
- The Apache Kafka broker must be accessible from the service
- The Prometheus API must be accessible from the service

The required python packages are depicted in the `requirements.txt` file. 

## Configuration

A variety of variables are defined in the `settings.py`  file. The configuration that is used in the deployment phase (either as `ENV` variables or via an `.env` file in case of docker-compose) includes:

| **Setting** | **Description** |
| --- | --- |
| DEBUG | Run the service on debug mode or not. By default, debug is disabled. |
| KAFKA_IP | The host of the Service Platform Virtualization publish/subscribe broker. |
| KAFKA_PORT | The port of the Service Platform Virtualization publish/subscribe broker. By default, port is 9092. |
| KAFKA_KUBERNETES_TOPIC | The publish/subscribe broker topic name where the monitoring data are published. By default, the topic name is `"nfvi.ncsrd.kubernetes"`. | 
| PROMETHEUS_HOST | The host of the Prometheus API hosted in the Kubernetes cluster. | 
| PROMETHEUS_PORT | The port of the Prometheus API hosted in the Kubernetes cluster. | 
| PROMETHEUS_POLLING_STEP | Describes the polling step in Prometheus API. Default value is `20s`. | 
| SCHEDULER_SECONDS | How frequent the publisher collects the monitoring data through the Prometheus API and publishes them in the pub/sub broker. Default value is `20` seconds. | 

## Installation/Deployment

To build the docker image, copy the bash script included in the `bash_scripts/` folder in the parent folder of the project and then, run:
```bash
    cd $HOME
    # clone repository
    cp ./kubernetes-prometheus-publisher/bash_scripts/build_docker_image.sh .
    chmod +x build_docker_image.sh
    ./build_docker_image.sh
```

Check the available docker images using:
```bash
    docker images
```

Considering the docker image is available, you can deploy the service as a docker container using the below command:
```bash
$ sudo docker run -p 80:3333 --name k8s-prometheus-publisher --restart always \
  -e DEBUG=1 \
  -e KAFKA_IP="192.168.1.175" \
  -e KAFKA_PORT="9092" \
  -e KAFKA_KUBERNETES_TOPIC="nfvi.ncsrd2.kubernetes" \
  -e PROMETHEUS_HOST="192.168.1.107" \
  -e PROMETHEUS_PORT="31078" \
  -e PROMETHEUS_POLLING_STEP="20s" \
  -e SCHEDULER_SECONDS=20 \
  -dit k8s-prometheus-publisher
```

Check the running docker containers:
```bash
    docker ps -a
```
The name of the docker image and container is:  *k8s-prometheus-publisher*.

Check the logs using the cmd:
```bash
    docker logs -f k8s-prometheus-publisher
    #expected logs
    Starting supervisor: supervisord.
    supervisord is running
    Initialization completed.
```


## Usage

After the installation/deployment of this service, it collects the
values of predefined set of metrics (as these defined in the
`settings.py` file) every `X` seconds and send them in the topic
`nfvi.ncsrd.kubernetes`.

An indicative structure of each message is:
```json
{
    "container_id": "959a1de912ce454a9a1de912ce654ae9",
    "metric_type": "container_cpu_system_seconds_total",
    "data": [
        {
            "type": "counter",
            "name": "container_cpu_system_seconds_total",
            "timestamp": "2019-02-13T10:14:32.897000Z",
            "unit": "seconds",
            "value": "0.21000000000000016"
        },
        {
            "type": "counter",
            "name": "container_cpu_system_seconds_total",
            "timestamp": "2019-02-13T10:15:02.897000Z",
            "unit": "seconds",
            "value": "0.21000000000000016"
        },
        {
            "type": "counter",
            "name": "container_cpu_system_seconds_total",
            "timestamp": "2019-02-13T10:15:32.897000Z",
            "unit": "seconds",
            "value": "0.21000000000000016"
        }
    ]
}
```

Considering that the service is running as a supervisor task, you can check its status either by typing in your browser `http://{host}` and using as username/password the admin/admin, or inspecting the `supervisorctl` inside the docker container:
```bash
    # supervisorctl
    k8s-prometheus-publisher         RUNNING   pid 25, uptime 0:01:30
```


## Tests

The entrypoint of the service is the `worker.py` file.

Therefore, you can type the below command to test it without the usage of the docker service after the installation of the needed packages (`pip3 install -r requirements.txt`):
```bash
$ python3 daemon.py
```

## Authors
- Singular Logic <pathanasoulis@ep.singularlogic.eu>

## Acknowledgements
This project has received funding from the European Union’s Horizon 2020 research and innovation programme under grant agreement [No 761699](http://www.5gmedia.eu/). The dissemination of results herein reflects only the author’s view and the European Commission is not responsible for any use that may be made 
of the information it contains.

## License
[Apache 2.0](LICENSE.md)


CHANGELOG.md

2019-10-03
-------------
- Provide documentation using sphinx (html, tex)

2019-02-20
-------------
- Support environmental variables (see `README`)

2018-12-20
-------------
- Update the query in Prometheus API considering the addition of the label `label_vim_id`
- Integration with Kafka bus

2018-12-19
-------------
- Fetch data from the Prometheus API using a scheduler

2018-12-18
-------------
- Initial contribution

utils.py
"""
A module that includes a set of utilities.
"""

from datetime import datetime, timedelta
import urllib.parse
from dateutil import tz
from settings import PROMETHEUS_METRICS_LIST, SCHEDULER_SECONDS, PROMETHEUS_POLLING_STEP


def convert_unix_timestamp_to_datetime_str(unix_ts):
    """ Convert a unix timestamp in stringify datetime (UTC)

    Args:
        unix_ts (int): The timestamp in unix

    Returns:
        str: The datetime in str (UTC)

    Example:
        >>> from datetime import datetime
        >>> from utils import convert_unix_timestamp_to_datetime_str
        >>> unix_ts = 1527165350
        >>> dt_str = convert_unix_timestamp_to_datetime_str(unix_ts)
        >>> print(dt_str)
        2018-05-24T12:35:50.000000Z
    """
    timestamp = datetime.utcfromtimestamp(unix_ts)
    return timestamp.strftime('%Y-%m-%dT%H:%M:%S.%fZ')


def convert_utc_timestamp_in_timezone(utc_timestamp, timezone="GMT"):
    """Convert the UTC timestamp to timezone by given new timezone.

        Args:
            utc_timestamp (object): A utc = datetime.utcnow() object
            timezone (str): A timezone i.e. Europe/Athens

        Returns:
            str: the timestamp in the given timezone considering the UTC timestamp
    """
    from_zone = tz.gettz('UTC')
    to_zone = tz.gettz(timezone)
    utc = utc_timestamp.replace(tzinfo=from_zone)
    return utc.astimezone(to_zone)


def apply_function_per_metric(metric_type):
    """ Select the proper function the applied in the web service that retrieves the values per pod.

    Args:
        metric_type (str): The type of the metric

    Returns:
        str: the name of the function
    """
    net_metrics = ["container_network_receive_bytes_total",
                   "container_network_receive_errors_total",
                   "container_network_receive_packets_dropped_total",
                   "container_network_receive_packets_total",
                   "container_network_transmit_bytes_total",
                   "container_network_transmit_errors_total",
                   "container_network_transmit_packets_dropped_total",
                   "container_network_transmit_packets_total", ]

    computational_metrics = ["container_cpu_usage_seconds_total",
                             "container_cpu_user_seconds_total",
                             "container_cpu_system_seconds_total ",
                             "container_cpu_cfs_throttled_seconds_total",
                             "container_fs_writes_bytes_total",
                             "container_fs_reads_bytes_total", ]

    if metric_type in net_metrics or metric_type in computational_metrics:
        return "rate"
    return "avg_over_time"


def retrieve_values(query, metric):
    """ Retrieve the values by given metric

    Args:
        query (object): The QueryRange object
        metric (str): The name of the metric

    Returns:
        object: a requests object
    """
    utc_now = datetime.utcnow()

    query = """sum(
      max(kube_pod_labels{label_ow_action!=""}) by (label_ow_action, pod, label_vim_id)
      *
      on(pod)
      group_right(label_ow_action, label_vim_id)
      label_replace(
        sum by (pod_name) (%(metric_function)s(%(metric_name)s{namespace="%(namespace)s"}[1m])), 
        "pod", 
        "$1", 
        "pod_name", 
        "(.+)"
      )
    ) by (pod, label_ow_action, label_vim_id)""" \
            % {"metric_function": apply_function_per_metric(metric), "metric_name": metric,
               "namespace": "default"}

    url_query = urllib.parse.quote(query)
    from_dt = utc_now - timedelta(seconds=int(SCHEDULER_SECONDS))
    from_time = from_dt.strftime('%Y-%m-%dT%H:%M:%S.%fZ')
    to_time = utc_now.strftime('%Y-%m-%dT%H:%M:%S.%fZ')
    return query.get(url_query, from_time=from_time, to_time=to_time, step=PROMETHEUS_POLLING_STEP)


def calculate_packet_loss_values(dropped_packets_rate, total_packets_rate):
    """ Calculate the packet loss (percentage)

    Args:
        dropped_packets_rate (float): The rate of dropped packets (packets/sec)
        total_packets_rate (float): The rate of total packets (packets/sec)

    Returns:
        float: The packet loss. Value range: [0 - 100]. Unit is %.
    """
    if float(total_packets_rate) == 0.0:
        return 0.0
    return float(dropped_packets_rate) * 100 / float(total_packets_rate)


def get_unit_by_metric(metric):
    """ Get the unit by given metric

    Args:
        metric (str): The metric name

    Returns:
        str: the unit
    """
    for item in PROMETHEUS_METRICS_LIST:
        if item['name'] == metric:
            return item['unit']
    return ""


def get_type_by_metric(metric):
    """ Get the type by given metric

    Args:
        metric (str): The metric name

    Returns:
        str: the metric's type
    """
    for item in PROMETHEUS_METRICS_LIST:
        if item['name'] == metric:
            return item['type']
    return ""


worker.py
"""
This module operates as the entry-point of this project.
"""

import json
import logging.config
import time
import schedule
from kafka import KafkaProducer
from kafka.errors import KafkaError
from prometheus_client.v1 import query_range
from settings import LOGGING, SCHEDULER_SECONDS, PROMETHEUS_METRICS_LIST, KAFKA_API_VERSION, \
    KAFKA_SERVER, KAFKA_KUBERNETES_TOPIC
from utils import convert_unix_timestamp_to_datetime_str, retrieve_values, \
    calculate_packet_loss_values

logging.config.dictConfig(LOGGING)
logger = logging.getLogger("publisher")
error_logger = logging.getLogger("errors")


def main():
    """main process"""
    producer = KafkaProducer(bootstrap_servers=KAFKA_SERVER,
                             api_version=KAFKA_API_VERSION,
                             value_serializer=lambda v: json.dumps(v).encode('utf-8'))
    prom_ql = query_range.QueryRange(token=None)

    # Keep metrics for the packet loss calculation
    tx_rx_metrics = {}

    for metric in PROMETHEUS_METRICS_LIST:
        try:
            # Load the values for the requested metric for any running container
            # having the label 'vim_id'
            response = retrieve_values(prom_ql, metric['name'])
            if response.status_code != 200:
                error_logger.error("GET {} - {}".format(response.url, response.text))
                continue

            response_body = response.json()
            result_type = response_body['data'].get('resultType')
            if result_type != "matrix":
                logger.warning("The `resultType` is not the matrix. It is `{}`".format(result_type))

            # The response of the `query_range` request returns the result type (`matrix`) and
            # a list of results. Each object in results list includes one or more values of the
            # requested metric per container ID; actually, the `label_vim_id` key reflects
            # the container ID.
            for result in response_body['data'].get('result', []):
                metric_values = []
                osm_container_id = result.get('metric', {}).get('label_vim_id', None)
                # Skip process if the container is not relevant with OSM
                if osm_container_id is None:
                    continue

                # Add an empty dict for an OSM container
                if osm_container_id not in tx_rx_metrics.keys():
                    tx_rx_metrics[osm_container_id] = {}

                # Keep all the values for the requested metric by given container ID in a list.
                values_len = len(result['values'])
                if values_len:
                    latest_value = result['values'][values_len - 1]
                    metric_values.append(
                        {"timestamp": convert_unix_timestamp_to_datetime_str(latest_value[0]),
                         "unit": metric['unit'], "type": metric['type'], "name": metric['name'],
                         "value": latest_value[1]})

                    # Save temporary a set of useful metrics for the packet loss calculation
                    if metric["name"] in ["container_network_receive_packets_dropped_total",
                                          "container_network_receive_packets_total",
                                          "container_network_transmit_packets_dropped_total",
                                          "container_network_transmit_packets_total"]:
                        tx_rx_metrics[osm_container_id][metric["name"]] = latest_value[1]
                        proper_tm = convert_unix_timestamp_to_datetime_str(latest_value[0])
                        tx_rx_metrics[osm_container_id]['timestamp'] = proper_tm

                    # Push the metric values in batch per container ID
                    payload = {"container_id": osm_container_id, "type": metric['type'],
                               "data": metric_values}
                    logger.debug("Generic metrics: {}".format(payload))
                    publish_metrics(producer, payload)
        except Exception as ex:
            error_logger.exception(ex)

    # The tx_rx_metrics keeps the required metrics for the calculation of the
    # packet loss in each container. Indicative sample:
    # --------------------
    # {
    #   "bfbcd872d6b64adcbcd872d6b64adcf1": {
    #     "timestamp": "2019-04-24T07:56:34.158000Z",
    #     "container_network_receive_packets_dropped_total": "0",
    #     "container_network_transmit_packets_dropped_total": "0",
    #     "container_network_transmit_packets_total": "15.5",
    #     "container_network_receive_packets_total": "36"
    #   },
    #   "c7a07d28d38746c6a07d28d38746c630": {
    #     "timestamp": "2019-04-24T07:56:34.158000Z",
    #     "container_network_receive_packets_dropped_total": "0",
    #     "container_network_transmit_packets_dropped_total": "0",
    #     "container_network_transmit_packets_total": "16.875",
    #     "container_network_receive_packets_total": "37.25"
    #   }
    # }
    for container in tx_rx_metrics.keys():
        # Packet loss in RX
        try:
            container_network_receive_packet_loss_percentage = calculate_packet_loss_values(
                tx_rx_metrics[container].get('container_network_receive_packets_dropped_total'),
                tx_rx_metrics[container].get('container_network_receive_packets_total'))
            rx_payload = {"container_id": container,
                          "type": 'container_network_receive_packet_loss_percentage',
                          "data": [{"timestamp": tx_rx_metrics[container]["timestamp"],
                                    "unit": "%", "type": "counter",
                                    "name": "container_network_receive_packet_loss_percentage",
                                    "value": container_network_receive_packet_loss_percentage}]}
            logger.debug("rx_packet_loss: {}".format(rx_payload))
            publish_metrics(producer, rx_payload)
        except TypeError as ex:
            error_logger.error(ex)

        # Packet loss in TX
        try:
            container_network_transmit_packet_loss_percentage = calculate_packet_loss_values(
                tx_rx_metrics[container].get('container_network_transmit_packets_dropped_total'),
                tx_rx_metrics[container].get('container_network_transmit_packets_total'))
            tx_payload = {"container_id": container,
                          "type": 'container_network_transmit_packet_loss_percentage',
                          "data": [{"timestamp": tx_rx_metrics[container]["timestamp"],
                                    "unit": "%", "type": "counter",
                                    "name": "container_network_transmit_packet_loss_percentage",
                                    "value": container_network_transmit_packet_loss_percentage}]}
            logger.debug("tx_packet_loss: {}".format(tx_payload))
            publish_metrics(producer, tx_payload)
        except TypeError as ex:
            error_logger.error(ex)

    # Close producer
    producer.close()


def publish_metrics(producer, payload):
    """ Publish the payload in kafka bus

    Args:
        producer (iterator): The kafka iterator
        payload (dict): The message to be published

    Returns:
        None
    """
    request = producer.send(KAFKA_KUBERNETES_TOPIC, payload)
    try:
        request.get(timeout=3)
    except KafkaError as ex:
        error_logger.error(ex)


if __name__ == '__main__':
    # Retrieve the data every X seconds
    schedule.every(int(SCHEDULER_SECONDS)).seconds.do(main)
    while True:
        schedule.run_pending()
        time.sleep(1)


deployment/run.sh
#!/bin/bash

# set the variables in the supervisor environment
sed -i "s/ENV_DEBUG/$DEBUG/g" /etc/supervisor/supervisord.conf
sed -i "s/ENV_KAFKA_IP/$KAFKA_IP/g" /etc/supervisor/supervisord.conf
sed -i "s/ENV_KAFKA_PORT/$KAFKA_PORT/g" /etc/supervisor/supervisord.conf
sed -i "s/ENV_KAFKA_KUBERNETES_TOPIC/$KAFKA_KUBERNETES_TOPIC/g" /etc/supervisor/supervisord.conf
sed -i "s/ENV_PROMETHEUS_HOST/$PROMETHEUS_HOST/g" /etc/supervisor/supervisord.conf
sed -i "s/ENV_PROMETHEUS_PORT/$PROMETHEUS_PORT/g" /etc/supervisor/supervisord.conf
sed -i "s/ENV_PROMETHEUS_POLLING_STEP/$PROMETHEUS_POLLING_STEP/g" /etc/supervisor/supervisord.conf
sed -i "s/ENV_SCHEDULER_SECONDS/$SCHEDULER_SECONDS/g" /etc/supervisor/supervisord.conf

# Restart services
service supervisor start && service supervisor status

# Makes services start on system start
update-rc.d supervisor defaults

echo "Initialization completed."
tail -f /dev/null  # Necessary in order for the container to not stop


prometheus_client/v1/query_range.py
"""
Module that implements the QueryRange requests in Prometheus API
"""

import logging.config

import urllib3

from httpclient.client import Client
from settings import PROMETHEUS, LOGGING

urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
logging.config.dictConfig(LOGGING)
logger = logging.getLogger("publisher")


class QueryRange(object):
    """QueryRange Class.

    Attributes:
        bearer_token (str, optional): The Prometheus Authorization Token (if any)

    Methods:
        get(query, from_time, to_time, step): perform a query_range request in Prometheus API
    """

    def __init__(self, token=None):
        """Class Constructor."""
        self.__client = Client(verify_ssl_cert=False)
        self.bearer_token = token

    def get(self, query, from_time, to_time, step=14):
        """ Perform a query_range request in Prometheus API (PromQL)

        Args:
            query (str): The query
            from_time (str): The start datetime
            to_time (str): The end datetime
            step (int): The used step in the query. Default value is 14 secs.

        Returns:
            object: A list of NSs as a requests object
        """
        endpoint = 'http://{}:{}/api/v1/query_range'.format(PROMETHEUS.get('HOST'),
                                                            PROMETHEUS.get('PORT'))
        headers = {"Accept": "application/json"}
        endpoint += "?query={}&start={}&end={}&step={}".format(query, from_time, to_time, step)
        logger.debug("Prometheus web service: {}".format(endpoint))
        response = self.__client.get(url=endpoint, headers=headers)
        return response


settings.py
"""
Definition of the settings
"""

import os

DEBUG = int(os.environ.get("DEBUG", 0))
PROJECT_ROOT = os.path.dirname(os.path.realpath(__file__))

# =================================
# KAFKA SETTINGS
# =================================
KAFKA_SERVER = "{}:{}".format(os.environ.get("KAFKA_IP", "192.168.1.175"),
                              os.environ.get("KAFKA_PORT", "9092"))
KAFKA_CLIENT_ID = 'kubernetes-prometheus-publisher'
KAFKA_API_VERSION = (1, 1, 0)
KAFKA_KUBERNETES_TOPIC = os.environ.get("KAFKA_KUBERNETES_TOPIC", "nfvi.ncsrd.kubernetes")

# =================================
# PROMETHEUS SETTINGS
# =================================
PROMETHEUS = {"HOST": os.environ.get("PROMETHEUS_HOST", "10.100.176.57"),
              "PORT": os.environ.get("PROMETHEUS_PORT", 31078)}
PROMETHEUS_POLLING_STEP = os.environ.get("PROMETHEUS_POLLING_STEP", '20s')
PROMETHEUS_METRICS_LIST = [
    {"name": "container_fs_inodes_free", "type": "gauge", "unit": ""},
    {"name": "container_fs_io_current", "type": "gauge", "unit": "iops"},
    {"name": "container_fs_usage_bytes", "type": "gauge", "unit": "bytes"},
    {"name": "container_fs_writes_bytes_total", "type": "counter", "unit": "bytes"},
    {"name": "container_fs_reads_bytes_total", "type": "counter", "unit": "bytes"},
    {"name": "container_cpu_load_average_10s", "type": "gauge", "unit": "seconds"},
    {"name": "container_cpu_system_seconds_total", "type": "counter", "unit": "seconds"},
    {"name": "container_cpu_usage_seconds_total", "type": "counter", "unit": "seconds"},
    {"name": "container_cpu_cfs_throttled_seconds_total", "type": "counter", "unit": "seconds"},
    {"name": "container_memory_usage_bytes", "type": "gauge", "unit": "bytes"},
    {"name": "container_memory_working_set_bytes", "type": "gauge", "unit": "bytes"},
    {"name": "container_memory_max_usage_bytes", "type": "gauge", "unit": "bytes"},
    {"name": "container_memory_swap", "type": "gauge", "unit": "bytes"},
    {"name": "container_network_receive_bytes_total", "type": "counter", "unit": "bytes"},
    {"name": "container_network_receive_errors_total", "type": "counter", "unit": ""},
    {"name": "container_network_receive_packets_dropped_total", "type": "counter", "unit": ""},
    {"name": "container_network_receive_packets_total", "type": "counter", "unit": "packets"},
    {"name": "container_network_transmit_bytes_total", "type": "counter", "unit": "bytes"},
    {"name": "container_network_transmit_errors_total", "type": "counter", "unit": ""},
    {"name": "container_network_transmit_packets_dropped_total", "type": "counter", "unit": ""},
    {"name": "container_network_transmit_packets_total", "type": "counter", "unit": ""},
    {"name": "container_network_tcp_usage_total", "type": "gauge", "unit": ""},
    {"name": "container_network_udp_usage_total", "type": "gauge", "unit": ""},
    {"name": "container_spec_memory_reservation_limit_bytes", "type": "gauge", "unit": "bytes"},
    {"name": "container_network_receive_packet_loss_percentage", "type": "counter", "unit": "%"},
    {"name": "container_network_transmit_packet_loss_percentage", "type": "counter", "unit": "%"}
]

# =================================
# SCHEDULER SETTINGS
# =================================
SCHEDULER_SECONDS = os.environ.get("SCHEDULER_SECONDS", 20)  # seconds

# ==================================
# LOGGING SETTINGS
# ==================================
# See more: https://docs.python.org/3.5/library/logging.config.html
LOGGING = {
    'version': 1,
    'disable_existing_loggers': True,
    'formatters': {
        'detailed': {
            'class': 'logging.Formatter',
            'format': "[%(asctime)s] - [%(name)s:%(lineno)s] - [%(levelname)s] %(message)s",
        },
        'simple': {
            'class': 'logging.Formatter',
            'format': '%(name)-15s %(levelname)-8s %(processName)-10s %(message)s'
        }
    },
    'handlers': {
        'console': {
            'class': 'logging.StreamHandler',
            'level': 'INFO',
            'formatter': 'simple',
        },
        'publisher': {
            'class': 'logging.handlers.RotatingFileHandler',
            'filename': "{}/logs/publisher.log".format(PROJECT_ROOT),
            'mode': 'w',
            'formatter': 'detailed',
            'level': 'DEBUG' if DEBUG else 'INFO',
            'maxBytes': 1024 * 1024,
            'backupCount': 5,
        },
        'errors': {
            'class': 'logging.handlers.RotatingFileHandler',
            'filename': "{}/logs/error.log".format(PROJECT_ROOT),
            'mode': 'w',
            'level': 'ERROR',
            'formatter': 'detailed',
            'maxBytes': 1024 * 1024,
            'backupCount': 5,
        },
    },
    'loggers': {
        'publisher': {
            'level': 'DEBUG',
            'handlers': ['publisher']
        },
        'errors': {
            'handlers': ['errors']
        }
    },
    'root': {
        'level': 'DEBUG',
        'handlers': ['console']
    }
}
